packages/aws-client/src/S3.ts

```diff
@@ -779,39 +706,148 @@ export const uploadS3FileStream = (
   }
   return contents;
 };
 
+/**
+ * Fetch complete list of S3 objects
+ *
+ * listObjectsV2 is limited to 1,000 results per call.  This function continues
+ * listing objects until there are no more to be fetched.
+ *
+ * The passed params must be compatible with the listObjectsV2 call.
+ *
+ * https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#listObjectsV2-property
+ *
+ * @param {Object} params - params for the s3.listObjectsV2 call
+ * @returns {Promise<Array>} resolves to an array of objects corresponding to
+ *   the Contents property of the listObjectsV2 response
+ *
+ * @static
+ */
+export const listS3ObjectsV2 = async (
+  params: ListObjectsV2Request
+): Promise<Array<ListS3ObjectsOutput>> => {
+  // Fetch the first list of objects from S3
+  let listObjectsResponse = await s3().listObjectsV2(params);
+
+  let discoveredObjects = listObjectsResponse.Contents ?? [];
+
+  // Keep listing more objects from S3 until we have all of them
+  while (listObjectsResponse.IsTruncated) {
+    // eslint-disable-next-line no-await-in-loop
+    listObjectsResponse = (await s3().listObjectsV2(
+      // Update the params with a Continuation Token
+      {
+
+        ...params,
+        ContinuationToken: listObjectsResponse.NextContinuationToken,
+      }
+    ));
+    discoveredObjects = discoveredObjects.concat(listObjectsResponse.Contents ?? []);
+  }
+
+  return discoveredObjects.filter((obj) => obj.Key) as Array<ListS3ObjectsOutput>;
+};
+
+/**
+ * Fetch lazy list of S3 objects
+ *
+ * listObjectsV2 is limited to 1,000 results per call.  This function continues
+ * listing objects until there are no more to be fetched.
+ *
+ * The passed params must be compatible with the listObjectsV2 call.
+ *
+ * https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#listObjectsV2-property
+ *
+ * @param params - params for the s3.listObjectsV2 call
+ * @yields a series of objects corresponding to
+ *   the Contents property of the listObjectsV2 response
+ *   batched to allow processing of one chunk at a time
+ *
+ * @static
+ */
+export async function* listS3ObjectsV2Batch(
+  params: ListObjectsV2Request
+): AsyncIterable<Array<ListS3ObjectsOutput>> {
+  let listObjectsResponse = await s3().listObjectsV2(params);
+
+  let discoveredObjects = listObjectsResponse.Contents ?? [];
+  yield discoveredObjects.filter((obj) => 'Key' in obj) as Array<ListS3ObjectsOutput>;
+  // Keep listing more objects from S3 until we have all of them
+  while (listObjectsResponse.IsTruncated) {
+    // eslint-disable-next-line no-await-in-loop
+    listObjectsResponse = (await s3().listObjectsV2(
+      // Update the params with a Continuation Token
+      {
+
+        ...params,
+        ContinuationToken: listObjectsResponse.NextContinuationToken,
+      }
+    ));
+    discoveredObjects = listObjectsResponse.Contents ?? [];
+    yield discoveredObjects.filter((obj) => 'Key' in obj) as Array<ListS3ObjectsOutput>;
+  }
+}
```

Comment:
Good Method!👍
Same question about the filter.
Return type should be ListS3ObjectsOutput['Cotents']